File: 0001-0099/0010. Regular Expression Matching/0010. Regular Expression Matching.js
/**
 * @param {string} str
 * @param {string} pattern
 * @return {boolean}
 */ 
//  dp[i][j] means to index i of s and index j of p, whether it match or not.
//  1, If p.charAt(j) == s.charAt(i) :  dp[i][j] = dp[i-1][j-1];
//  2, If p.charAt(j) == '.' : dp[i][j] = dp[i-1][j-1];
//  3, If p.charAt(j) == '*': 
//     here are two sub conditions:
//                 1   if p.charAt(j-1) != s.charAt(i) : dp[i][j] = dp[i][j-2]  //in this case, a* only counts as empty
//                 2   if p.charAt(i-1) == s.charAt(i) or p.charAt(j-1) == '.':
//                                dp[i][j] = dp[i-1][j]    //in this case, a* counts as multiple a 
//                             or dp[i][j] = dp[i][j-1]   // in this case, a* counts as single a
//                             or dp[i][j] = dp[i][j-2]   // in this case, a* counts as empty
var isMatch = function (str, pattern) {
  const dp = Array.from({length:str.length+1},e=>Array(pattern.length+1).fill(false));
  console.log(dp)
  dp[0][0] = true;  // 占位符 不代表任何位置  代表str pattern都是空字符的时候 
  // 由于3.1只能处理非边界情况  如果pattern 以 a*开头 3.1的模式无法处理所以需要在这里做特殊处理
  for (let j = 0; i < pattern.length; j++)
    dp[0][j + 1] = pattern[i] === "*" && dp[0][j - 1];
    //如果开头是连续的 a*b* 由于x*可以代表0个x字符  那么应当永远和str是空字符匹配  当str时空的时候的情况
  // i代表str的索引   j代表pattern的索引  
  for (let i = 0; i < str.length; i++) {
    for (let j = 0; j < pattern.length; j++) {
      if (pattern[j] === ".") {
        dp[i + 1][j + 1] = dp[i][j];  //需要注意  dp[i+1][j+1]代表的 str[i]和pattern[j]是否匹配 因为dp长宽比str和pattern都大一位
      }
      if (pattern[j] === str[i]) {
        dp[i + 1][j + 1] = dp[i][j];
      }
      
      if (pattern[j] === "*") {
        if (pattern[j - 1] !== str[i] && pattern[j - 1] !== ".") {  //此时 * 匹配0个元素
          dp[i + 1][j + 1] = dp[i + 1][j - 1];  // str[i] pattern[j]的匹配情况和str[i] pattern[j-2]的匹配情况一样
        } else {
          dp[i + 1][j + 1] = dp[i + 1][j] ||   //匹配单个a  或者单个任意字符
          dp[i][j + 1] ||  // 多个a  或者多个任意字符
          //pattern[j]=== "*" && str[i] === pattern[j-1] && (str[i]和pattern[j-1]不匹配),说明有多个a才不匹配,不然肯定匹配,因为pattern[j-1]必定是个字母或者. ,去掉 * 就是单个字母,如果不匹配字母又相等 那说明  str = "xaa"  pattern="xa*" "xa"无法匹配"xaa"
          dp[i + 1][j - 1]; // empty 匹配   aasdf  aasdf.*
          // 为星号要么匹配0个要么匹配1个要么匹配多个
          // 前面判断过  x* 匹配0次这种模式了 
          // 还剩下几种情况
          // xa  xa*  x.*  匹配一次
          // xaa xa*  x.* 匹配多次
          // x   x.* 匹配0次 前面检测过 不为. 这里只需要检测 .*
         }
      }
    }
  }
  return dp[str.length][pattern.length];
};
isMatch("aab", "c*a*b")

File: 0001-0099/0010. Regular Expression Matching/0010. Regular Expression Matching.py
# 1, If p.charAt(j) == s.charAt(i) :  dp[i][j] = dp[i-1][j-1];
# 2, If p.charAt(j) == '.' : dp[i][j] = dp[i-1][j-1];
# 3, If p.charAt(j) == '*':
#    here are two sub conditions:
#                1   if p.charAt(j-1) != s.charAt(i) : dp[i][j] = dp[i][j-2]  //in this case, a* only counts as empty
#                2   if p.charAt(i-1) == s.charAt(i) or p.charAt(i-1) == '.':
#                               dp[i][j] = dp[i-1][j]    //in this case, a* counts as multiple a
#                            or dp[i][j] = dp[i][j-1]   // in this case, a* counts as single a
#                            or dp[i][j] = dp[i][j-2]   // in this case, a* counts as empty
from functools import cache
import numpy


class Solution:

    def isMatch(self, s: str, p: str) -> bool:
        dp = [[False] * (len(p) + 1) for _ in range(len(s) + 1)]
        dp[0][0] = True
        for j in range(len(p)):  # 1
            dp[0][j + 1] = p[j] == '*' and dp[0][j - 1]
        for i in range(len(s)):
            for j in range(len(p)):
                if p[j] == '.':
                    dp[i + 1][j + 1] = dp[i][j]
                if p[j] == s[i]:
                    dp[i + 1][j + 1] = dp[i][j]
                if p[j] == '*':
                    if p[j - 1] != s[i] and p[j - 1] != '.':
                        dp[i + 1][j + 1] = dp[i + 1][j - 1]
                    else:
                        dp[i + 1][j + 1] = dp[i + 1][j] or dp[i][j + 1] or dp[i + 1][j - 1]
        print(numpy.matrix(dp))
        return dp[-1][-1]


class Solution:
    # TLE
    def isMatch2(self, text, pattern):
        if not pattern:
            return not text
        # bool(str) empty strings evaluate to False, but everything else evaluates to True
        first_match = bool(text) and pattern[0] in {text[0], '.'}
        if len(pattern) >= 2 and pattern[1] == '*':
            return self.isMatch2(text, pattern[2:]) or (first_match and self.isMatch2(text[1:], pattern))
        else:
            return first_match and self.isMatch2(text[1:], pattern[1:])

    def isMatch(self, text, pattern):

        @cache
        def dp(i, j):
            if j == len(pattern):
                ans = i == len(text)
            else:
                first_match = i < len(text) and pattern[j] in {text[i], '.'}
                if j + 1 < len(pattern) and pattern[j + 1] == '*':
                    ans = dp(i, j + 2) or first_match and dp(i + 1, j)
                else:
                    ans = first_match and dp(i + 1, j + 1)
            return ans

        return dp(0, 0)

    def isMatch3(self, s: str, p: str) -> bool:
        m, n = len(s), len(p)
        dp = [[False] * (n + 1) for _ in range(m + 1)]

        for i in range(m, -1, -1):
            for j in range(n, -1, -1):
                if i == m and j == n:
                    dp[i][j] = True
                else:
                    first_match = i < m and j < n and p[j] in {s[i], '.'}
                    if j + 1 < n and p[j + 1] == '*':
                        dp[i][j] = dp[i][j + 2] or first_match and dp[i + 1][j]
                    else:
                        dp[i][j] = first_match and dp[i + 1][j + 1]

        return dp[0][0]


# dp 代表的是字符的长度

so = Solution()
# so.isMatch('aab', 'c*a*b')
so.isMatch2("abc", "a.c")
# 1 处理这种情况
# s = "aab" p = "c*a*b"